sway-core/src/semantic_analysis/namespace/items.rs

```diff
@@ -101,31 +98,51 @@ impl Items {
         name: Ident,
         item: ty::TyDeclaration,
     ) -> CompileResult<()> {
-        let mut warnings = vec![];
         let mut errors = vec![];
-        // purposefully do not preemptively return errors so that the
-        // new definition allows later usages to compile
-        if self.symbols.get(&name).is_some() {
-            match item {
-                ty::TyDeclaration::EnumDeclaration { .. }
-                | ty::TyDeclaration::StructDeclaration { .. }
-                | ty::TyDeclaration::AbiDeclaration { .. }
-                | ty::TyDeclaration::TraitDeclaration { .. } => {
-                    errors.push(CompileError::ShadowsOtherSymbol { name: name.clone() });
-                }
-                ty::TyDeclaration::GenericTypeForFunctionScope { .. } => {
+
+        let append_shadowing_error =
+            |decl: &ty::TyDeclaration, item: &ty::TyDeclaration, errors: &mut Vec<CompileError>| {
+                use ty::TyDeclaration::*;
+                match (decl, &item) {
+                // variable shadowing a constant
+                // constant shadowing a constant
+                (
+                    ConstantDeclaration { .. },
+                    VariableDeclaration { .. } | ConstantDeclaration { .. },
+                )
+                // constant shadowing a variable
+                | (VariableDeclaration { .. }, ConstantDeclaration { .. })
+                // type shadowing another type
+                // trait/abi shadowing another trait/abi
+                // type shadowing a trait/abi or vice versa
+                | (
+                    StructDeclaration { .. }
+                    | EnumDeclaration { .. }
+                    | TraitDeclaration { .. }
+                    | AbiDeclaration { .. },
+                    StructDeclaration { .. }
+                    | EnumDeclaration { .. }
+                    | TraitDeclaration { .. }
+                    | AbiDeclaration { .. },
+                ) => errors.push(CompileError::NameDefinedMultipleTimes { name: name.to_string(), span: name.span() }),
```

Comment:
The formatter is not able to format this closure properly 🤷

Reply:
Good to know it's not just us struggling with formatter correctness :sweat_smile:

Reply:
I love how the pattern matching here clearly lays out our shadowing rules btw :heart: